mitodl/open-discussions
# upgrade flow to @latest

#### Pre-Flight checklist



#### What are the relevant tickets?

I was working on #1299 and I ran into an issue where a newer feature of React wasn't supported by the version of flow we are currently using, so I went ahead and upgraded it.

The newer version had a big change (see here: https://medium.com/flow-type/asking-for-required-annotations-64d4f9c1edf8) where in many cases type parameters that should have been required were not, and lots of typings were silently failing. 

Upgrading lets us also use newer libdefs for things like `react-redux` - in particular, with the new version of that `connect` is no longer a type black hole!

Now instead of just having the return value of `connect` be more or less totally untyped (as far as flow knows anyhow) it accepts a series of parameters, and it will actually check that props that the component is supposed to receive from parents are passed down.

To do this you need to declare new types, `StateProps`, `OwnProps`, and `DispatchProps`. `StateProps` is the return value of the `mapStateToProps` function, and includes any props that come from the redux state. `OwnProps` is props which the component receives from it's parents, and `DispatchProps` are the props which come from `mapDispatchToProps`. Then the type which the unconnected component itself uses is

```js
type Props = {|
  ...OwnProps,
  ...StateProps,
  ...DispatchProps
|}
```

then you annotate `connect` like

```js
export default connect<Props, OwnProps, _, DispatchProps, _,_>(mapStateToProps)(Component)
```

the `_` syntax tells Flow to infer those types (you can read the react-redux typedef docs to see more).

Anyway, this is really nice! Unfortunately it still isn't fully supported in the places where we use `R.compose` for this purpose, since the ramda libdef isn't compatible easily with our usage (I'm talking 7000 type errors, lol), but for more simple components it's a good thing imo.


#### What's this PR do?

upgrades to flow `0.94.0`, and fixes all the type errors that came up. I also bumped the libdefs for a few libraries up to the latest one (we might want to more comprehensively update those soon, but they can be a big pain).

#### How should this be manually tested?

The app should function as normal. There are very few actual code changes, the vast majority of the changes are just flow things, so I would say read through the code and make sure it makes sense, and where there is a code change make sure it's good. Also probably good to just do a basic smoke test of main functions of the app.
